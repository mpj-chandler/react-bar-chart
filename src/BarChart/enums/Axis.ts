enum Axis {
    XAxis,
    YAxis,
}

export default Axis;
