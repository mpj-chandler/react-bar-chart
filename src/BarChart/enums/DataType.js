"use strict";
exports.__esModule = true;
var DataType;
(function (DataType) {
    DataType["Named"] = "NAMED";
    DataType["DateIndexed"] = "DATE_INDEXED";
    DataType["Numeric"] = "NUMERIC";
    DataType["NonNullNumeric"] = "NON_NULL_NUMERIC";
})(DataType || (DataType = {}));
;
exports["default"] = DataType;
