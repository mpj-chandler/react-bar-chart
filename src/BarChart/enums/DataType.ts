enum DataType {
    Named = 'NAMED',
    DateIndexed = 'DATE_INDEXED',
    Numeric = 'NUMERIC',
    NonNullNumeric = 'NON_NULL_NUMERIC',
};

export default DataType;
