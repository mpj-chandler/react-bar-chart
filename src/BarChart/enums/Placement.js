"use strict";
exports.__esModule = true;
var Placement;
(function (Placement) {
    Placement[Placement["Aligned"] = 0] = "Aligned";
    Placement[Placement["Bucket"] = 1] = "Bucket";
})(Placement || (Placement = {}));
exports["default"] = Placement;
