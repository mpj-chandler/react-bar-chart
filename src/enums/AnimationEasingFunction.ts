enum AnimationEasingType {
    Linear = 'linear',
    Elastic = 'elastic',
    InExpo = 'inExpo',
    None = 'none'
}

export default AnimationEasingType;