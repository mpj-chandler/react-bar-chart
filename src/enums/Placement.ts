enum Placement {
    Aligned,
    Bucket,
}

export default Placement;
