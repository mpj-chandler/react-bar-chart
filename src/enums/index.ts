import Axis from './Axis';
import DataType from './DataType';
import { Placement } from './Placement';

export {
    Axis,
    DataType,
    Placement
}