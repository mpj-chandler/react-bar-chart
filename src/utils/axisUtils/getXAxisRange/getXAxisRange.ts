import Placement from '../../../enums/Placement';
import { applyAxisConfig } from '../applyAxisConfig/applyAxisConfig';
import { applyBucketPlacement } from '../applyBucketPlacement/applyBucketPlacement';
import extractMaximumAndMinimum from '../extractMaximumAndMinimum/extractMaximumAndMinimum';
import Axis from '../../../enums/Axis';
import { extractDataType } from '../generateAxisLabels/extractDataType';
import DataType from '../../../enums/DataType';
import { DataPoint, SeriesData, NamedDataPoint } from '../../../__types__/seriesTypes';
import { AxisConfig, AxisRange } from '../../../__types__/axisTypes';

export function getXAxisRange(seriesData: SeriesData[], config?: AxisConfig) {
    const dataType: DataType = extractDataType(seriesData, Axis.XAxis);

    switch(dataType) {
        case DataType.NonNullNumeric:
        case DataType.Numeric:
            return extractNumericXAxisRange(seriesData, config);
        case DataType.Named:
            return extractNamedXAxisRange(seriesData, config);
        default:
            throw new Error('Error in XAxis calibration - unrecognised data type!');
    }
}

function extractNamedXAxisRange(seriesData: SeriesData[], config?: AxisConfig) {
    const labels: Set<string> = new Set([]);

    seriesData.forEach((data: SeriesData) => {
        data.points.forEach((point: DataPoint) => {
            labels.add((point as NamedDataPoint).x);
        });
    });

    if (Array.from(labels).length === 0) {
        throw new Error('Error in XAxis calibration - unable to extract any data points!');
    }

    let range: AxisRange = {
        min: 0,
        max: Array.from(labels).length - 1,
    };

    if (config) {
        if (config.tickPlacement === Placement.Bucket) {
            range = applyBucketPlacement(range, seriesData[0].points.length);
        }
        range = applyAxisConfig(range, config);
    }

    return range;
}

function extractNumericXAxisRange(seriesData: SeriesData[], config?: AxisConfig) {
    let range: AxisRange = extractMaximumAndMinimum(seriesData, Axis.XAxis);
    if (config) {
        if (config.tickPlacement === Placement.Bucket) {
            range = applyBucketPlacement(range, seriesData[0].points.length);
        }
        range = applyAxisConfig(range, config);
    }

    return range;
}

