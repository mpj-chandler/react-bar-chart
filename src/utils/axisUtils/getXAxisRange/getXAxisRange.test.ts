import { getXAxisRange } from './getXAxisRange';
import { Placement } from '../../../enums/Placement';
import DataType from '../../../enums/DataType';

describe('getXAxisRange', () => {
    describe('when using non-null numeric data', () => {
        const data = [
            {
                seriesName: 'A',
                type: {
                    x: DataType.NonNullNumeric,
                    y: DataType.NonNullNumeric,
                },
                points: [
                    {
                        x: 0,
                        y: 10,
                    },
                    {
                        x: 1,
                        y: 20,
                    },
                    {
                        x: 2,
                        y: 30,
                    },
                ],
            },
        ];

        describe('When not provided with a config object', () => {
            it('it extracts the series maximum and minimum correctly', () => {
                const range = getXAxisRange(data);

                expect(range).toStrictEqual({ max: 2, min: 0 });
            });
        });

        describe('When provided with a config object', () => {
            it('it extracts the series maximum and minimum correctly', () => {
                const range = getXAxisRange(
                    data,
                    {
                        zeroIntercept: false,
                        margin: 400,
                        tickPlacement: Placement.Aligned,
                        tickLength: 10,
                    });

                expect(range).toStrictEqual({ max: 10, min: 0 });
            });
        });
    });

    describe('when using numeric data', () => {
        const data = [
            {
                seriesName: 'A',
                type: {
                    x: DataType.Numeric,
                    y: DataType.Numeric,
                },
                points: [
                    {
                        x: 0,
                        y: 10,
                    },
                    {
                        x: 1,
                        y: 20,
                    },
                    {
                        x: 2,
                        y: 30,
                    },
                ],
            },
        ];

        describe('When not provided with a config object', () => {
            it('it extracts the series maximum and minimum correctly', () => {
                const range = getXAxisRange(data);

                expect(range).toStrictEqual({ max: 2, min: 0 });
            });
        });

        describe('When provided with a config object', () => {
            it('it extracts the series maximum and minimum correctly', () => {
                const range = getXAxisRange(
                    data,
                    {
                        zeroIntercept: false,
                        margin: 400,
                        tickPlacement: Placement.Bucket,
                        tickLength: 10,
                    });

                expect(range).toStrictEqual({ max: 13.333333333333332, min: 0 });
            });
        });
    });

    describe('when using named data', () => {
        const data = [
            {
                seriesName: 'A',
                type: {
                    x: DataType.Named,
                    y: DataType.Numeric,
                },
                points: [
                    {
                        x: 'A',
                        y: 10,
                    },
                    {
                        x: 'B',
                        y: 20,
                    },
                    {
                        x: 'C',
                        y: 30,
                    },
                ],
            },
        ];

        describe('When not provided with a config object', () => {
            it('it extracts the series maximum and minimum correctly', () => {
                const range = getXAxisRange(data);

                expect(range).toStrictEqual({ max: 2, min: 0 });
            });
        });

        describe('When provided with a bucket config object', () => {
            it('it extracts the series maximum and minimum correctly', () => {
                const range = getXAxisRange(
                    data,
                    {
                        zeroIntercept: false,
                        margin: 400,
                        tickPlacement: Placement.Bucket,
                        tickLength: 10,
                    });

                expect(range).toStrictEqual({ max: 13.333333333333332, min: 0 });
            });
        });

        describe('When provided with an aligned config object', () => {
            it('it extracts the series maximum and minimum correctly', () => {
                const range = getXAxisRange(
                    data,
                    {
                        zeroIntercept: false,
                        margin: 400,
                        tickPlacement: Placement.Aligned,
                        tickLength: 10,
                    });

                expect(range).toStrictEqual({ max: 10, min: 0 });
            });
        });

        describe('When provided with empty data', () => {
            it('it throws an error', () => {
                const emptyData = [
                    {
                        seriesName: 'A',
                        type: {
                            x: DataType.Named,
                            y: DataType.Numeric,
                        },
                        points: [],
                    },
                ];
                expect(() => getXAxisRange(emptyData)).toThrowErrorMatchingSnapshot();
            });
        });
    });

    describe('when using an unknown data type', () => {
        const data = [
            {
                seriesName: 'A',
                type: {
                    x: DataType.DateIndexed,
                    y: DataType.Numeric,
                },
                points: [
                    {
                        x: 'A',
                        y: 10,
                    },
                    {
                        x: 'B',
                        y: 20,
                    },
                    {
                        x: 'C',
                        y: 30,
                    },
                ],
            },
        ];

        it('it throws an error', () => {
            expect(() => getXAxisRange(data)).toThrowErrorMatchingSnapshot();
        });
    });
});
