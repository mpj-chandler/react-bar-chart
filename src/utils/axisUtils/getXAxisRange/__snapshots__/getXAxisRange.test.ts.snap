// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getXAxisRange when using an unknown data type it throws an error 1`] = `"Error in XAxis calibration - unrecognised data type!"`;

exports[`getXAxisRange when using named data When provided with empty data it throws an error 1`] = `"Error in XAxis calibration - unable to extract any data points!"`;
